client/extensions/woocommerce/app/settings/email/mailchimp/setup-mailchimp.js

```diff
@@ -0,0 +1,315 @@
+		this.props.onClose();
+	}
+
+	getStoreSettings = () => {
+		// clear this and pass only what is required.
+		const { address, currency } = this.props;
+
+		const settings = pick( this.state.settings, storSettingsRequriredFields );
+		settings.store_city = address.city;
+		settings.store_street = address.street + ' ' + address.street2;
+		settings.store_state = address.state;
+		settings.store_country = address.country;
+		settings.store_postal_code = address.postcode;
+		settings.store_currency_code = currency;
+		return settings;
+	}
+
+	getCampaingDefaultsSettings = () => {
+		return pick( this.state.settings, campaignDefaultsRequiredFields );
+	}
+
+	hasEmptyValues = ( data ) => {
+		return some( data, isEmpty );
+	}
+
+	areStoreSettingsValid = ( settings ) => {
+		const hasAllKeys = storSettingsRequriredFields.every( key => key in settings );
+		if ( ! hasAllKeys ) {
+			return false;
+		}
+		const hasEmptyValues = this.hasEmptyValues( settings );
+		if ( hasEmptyValues ) {
+			return false;
+		}
+		if ( settings.store_phone.length <= 6 ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	areCampaignSettingsValid = ( settings ) => {
+		const hasAllKeys = campaignDefaultsRequiredFields.every( key => key in settings );
+		if ( ! hasAllKeys ) {
+			return false;
+		}
+		return true;
+	}
+
+	next = () => {
+		// Don't send data if it is not compleate or it is the same as the one we already had from
+		// the endpoint.
+		if ( this.state.step === KEY_INPUT_STEP ) {
+			this.props.submitMailChimpApiKey( this.props.siteId, this.state.api_key_input );
+			return;
+		} else if ( this.state.step === STORE_INFO_STEP ) {
+			const settings = this.getStoreSettings();
+			const validSettings = this.areStoreSettingsValid( settings );
+			if ( ! validSettings ) {
+				this.setState( { settings_values_missing: true } );
+				return;
+			}
+			this.setState( { settings_values_missing: false } );
+
+			this.props.submitMailchimpStoreInfo( this.props.siteId, settings );
+			return;
+		} else if ( this.state.step === CAMPAIGN_DEFAULTS_STEP ) {
+			const settings = this.getCampaingDefaultsSettings();
+			const validSettings = this.areCampaignSettingsValid( settings );
+			if ( ! validSettings ) {
+				this.setState( { settings_values_missing: true } );
+				return;
+			}
+			this.setState( { settings_values_missing: false } );
+			this.props.submitMailchimpCampaignDefaults( this.props.siteId, settings );
+			return;
+		} else if ( this.state.step === NEWSLETTER_SETTINGS_STEP ) {
+			const mailchimp_list = this.state.settings.mailchimp_list;
+			if ( ! mailchimp_list ) {
+				this.setState( { settings_values_missing: true } );
+				return;
+			}
+			this.setState( { settings_values_missing: false } );
+			this.props.submitMailchimpNewsletterSettings( this.props.siteId, { mailchimp_list } );
+			return;
+		}
+		this.setState( { step: steps[ this.state.step ].nextStep } );
+	}
+
+	onKeyInputChange = ( e ) => {
+		this.setState( { api_key_input: e.target.value } );
```

Comment:
currently won't this trigger a re-render on each key-up in the API key field? Granted most people will likely paste into this field - maybe we should debounce this, or have some sort of a 'save' button instead?